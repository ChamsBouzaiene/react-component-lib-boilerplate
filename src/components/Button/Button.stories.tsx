import React from 'react';
import { Story, Meta } from '@storybook/react';

import Button, { ButtonProps } from './Button';

export default {
    title: 'Components/Button',
    component: Button,
    argTypes: {
        backgroundColor: { control: 'color' },
        size: {
            control: {
                type: 'select',
                options: ['small', 'medium', 'large'],
            },
        },
        borderRadius: {
            control: {
                type: 'select',
                options: [2, 4, 8],
            },
        },
    },
} as Meta;

const Template: Story<ButtonProps> = (args) => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
    variant: 'primary',
    label: 'Button',
};

export const Secondary = Template.bind({});
Secondary.args = {
    variant: 'secondary',
    label: 'Button',
};

export const Danger = Template.bind({});
Danger.args = {
    variant: 'danger',
    label: 'Button',
};

export const Large = Template.bind({});
Large.args = {
    size: 'large',
    label: 'Button',
};

export const Medium = Template.bind({});
Medium.args = {
    size: 'medium',
    label: 'Button',
};

export const Small = Template.bind({});
Small.args = {
    size: 'small',
    label: 'Button',
};

export const Rad2 = Template.bind({});
Rad2.args = {
    borderRadius: 2,
    label: 'Button',
};

export const Rad4 = Template.bind({});
Rad4.args = {
    borderRadius: 4,
    label: 'Button',
};

export const Rad8 = Template.bind({});
Rad8.args = {
    borderRadius: 8,
    label: 'Button',
};
