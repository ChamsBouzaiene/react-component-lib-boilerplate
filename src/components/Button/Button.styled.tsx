import React from 'react';
import styled from 'styled-components';
import {
    borderRadius,
    buttonStyle,
    color,
    ColorProps,
    variant,
    space,
    fontSize,
    SpaceProps,
    FontSizeProps,
} from 'styled-system';

interface DefaultProps {
    borderRadius?: 0 | 2 | 4 | 8;
    variant?: 'primary' | 'secondary' | 'warning' | 'danger' | 'white';
    size?: 'small' | 'medium' | 'large';
}
interface Props extends ColorProps, DefaultProps, SpaceProps, FontSizeProps {
    children: React.ReactNode;
}

const buttonSize = variant({
    prop: 'size',
    key: 'buttonSizes',
});

export const ThemedButton = styled('button')<Props>(
    {
        appearance: 'button',
        border: 0,
        outline: 0,
    },
    borderRadius,
    buttonSize,
    buttonStyle,
    color,
    space,
    fontSize,
);
