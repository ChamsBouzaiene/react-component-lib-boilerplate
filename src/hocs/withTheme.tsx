import React from 'react';
import { ThemeProvider } from 'styled-components';
import darkTheme from '../theme';

export function withTheme<P>(WrappedComponent: React.ComponentType<P>): React.FC<P> {
    const ComponentWithTheme = (props: P) => {
        return (
            <ThemeProvider theme={darkTheme}>
                <WrappedComponent {...props} />
            </ThemeProvider>
        );
    };
    return ComponentWithTheme;
}
export default withTheme;
