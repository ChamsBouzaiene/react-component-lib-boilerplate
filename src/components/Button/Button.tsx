import React from 'react';
import { ThemedButton } from './Button.styled';
import { SpaceProps, FontSizeProps } from 'styled-system';
import PropTypes from 'prop-types';

export interface ButtonProps extends SpaceProps, FontSizeProps {
    /**
     * Is this the principal call to action on the page?
     */

    borderRadius?: 0 | 2 | 4 | 8;

    variant?: 'primary' | 'secondary' | 'warning' | 'danger' | 'white';
    /**
     * What background color to use
     */
    backgroundColor?: string;
    /**
     * How large should the button be?
     */
    size?: 'small' | 'medium' | 'large';
    /**
     * Button contents
     */
    label: string;
    /**
     * Optional click handler
     */
    onClick?: () => void;
}

/**
 * Button UI component for user interaction
 */
const Button: React.FC<ButtonProps> = ({ variant, size, label, ...props }) => (
    <ThemedButton variant={variant} size={size} {...props}>
        {label}
    </ThemedButton>
);

Button.defaultProps = {
    variant: 'primary',
    size: 'small',
    borderRadius: 2,
};

Button.propTypes = {
    borderRadius: PropTypes.oneOf([0, 2, 4, 8]),
    variant: PropTypes.oneOf(['primary', 'secondary', 'warning', 'danger', 'white']),
    backgroundColor: PropTypes.string,
    size: PropTypes.oneOf(['small', 'medium', 'large']),
    label: PropTypes.string.isRequired,
    onClick: PropTypes.func,
};

export default Button;
