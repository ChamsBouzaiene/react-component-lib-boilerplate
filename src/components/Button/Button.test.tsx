import React from 'react';
import { render, screen } from '@testing-library/react';
import Button from './Button';
import { withTheme } from '../../hocs/withTheme';
import userEvent from '@testing-library/user-event';

const ThemedButton = withTheme(Button);
describe('Button', () => {
    test('renders a default button with text', () => {
        const { baseElement } = render(<ThemedButton label={'Click me'} />);

        expect(screen.getByText('Click me')).toBeInTheDocument();
        expect(screen.getByText('Click me')).toHaveStyle({
            backgroundColor: '#025D92',
            color: 'white',
        });
        expect(baseElement).toMatchSnapshot();
    });
    test('renders a primary button', () => {
        const { baseElement } = render(<ThemedButton variant="primary" label={'Click me'} />);

        expect(screen.getByText('Click me')).toHaveStyle({
            backgroundColor: '#025D92',
            color: 'white',
        });
        expect(baseElement).toMatchSnapshot();
    });
    test('handles onClick', () => {
        const mockOnClick = jest.fn();
        const { baseElement } = render(<ThemedButton label={'Click Me'} onClick={mockOnClick} />);
        userEvent.click(screen.getByText('Click Me'));

        expect(mockOnClick).toHaveBeenCalledTimes(1);
        expect(baseElement).toMatchSnapshot();
    });
});
