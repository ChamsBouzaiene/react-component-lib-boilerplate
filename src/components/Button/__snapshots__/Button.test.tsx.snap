// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button handles onClick 1`] = `
<body>
  <div>
    <button
      class="sc-bdnxRM dTEMSW"
    >
      Click Me
    </button>
  </div>
</body>
`;

exports[`Button renders a default button with text 1`] = `
<body>
  <div>
    <button
      class="sc-bdnxRM dTEMSW"
    >
      Click me
    </button>
  </div>
</body>
`;

exports[`Button renders a primary button 1`] = `
<body>
  <div>
    <button
      class="sc-bdnxRM dTEMSW"
    >
      Click me
    </button>
  </div>
</body>
`;
